### ISSUE TITLE
how to add journals to doi-utils

### ISSUE THREAD

**Issue by brittAnderson** [jkitchin/org-ref#112]
Is there any benefit to trying to make this a bit more generic since many journal share a basic format, i.e. http(s)://<journalname>.(org,com)/(content,"").(full.)pdf ? Or is it just as easy to keep adding journalname-pdf-url functions for each one as needed? I don't mind doing the latter if that works for you.

**Comment by jkitchin** [comment]
good idea. I tried adding a generic function that checks if the .full.pdf url exists. could you try it?

**Comment by brittAnderson** [comment]
Unfortunately, I can't get that far. Before using the git version, I updated the elpa version and it is broken.  org-ref-20160120.856 `doi-utils-insert-bibtex-entry-from-doi` fails with wrong argument stringp nil . 

I also notice that it tries to insert the bibtex entry into the current buffer I have opened. Maybe something is left over from recent edits or debuggin?

Will try as soon as I can get it operating again.

**Comment by jkitchin** [comment]
what doi did it fail with? I just tried it with a fresh version from melpa and it seemed to work ok.

**Comment by brittAnderson** [comment]
I am baffled. I can no longer get org-ref to insert into bibtex from doi and I have absolutely no idea what has changed on my system (two machines). I went so far as to reinstall emacs and my elpa folder. and still the same error. You will see the doi in the backtrace, and the doi does work at dx.doi.org. Here is the backtrace:

Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  looking-at(nil)
  bibtex-print-help-message("title")
  bibtex-find-text(nil nil t)
  bibtex-next-field(nil)
  bibtex-set-field("DATE_ADDED" "Wed Jan 20 22:01:43 2016")
  doi-utils-insert-bibtex-entry-from-doi("10.1073/pnas.1517441113")
  funcall-interactively(doi-utils-insert-bibtex-entry-from-doi "10.1073/pnas.1517441113")
  call-interactively(doi-utils-insert-bibtex-entry-from-doi record nil)
  command-execute(doi-utils-insert-bibtex-entry-from-doi record)
  helm-M-x(nil "doi-utils-insert-bibtex-entry-from-doi")
  funcall-interactively(helm-M-x nil "doi-utils-insert-bibtex-entry-from-doi")
  call-interactively(helm-M-x nil nil)
  command-execute(helm-M-x)

**Comment by jkitchin** [comment]
I am a bit baffled too. I will check it out tomorrow afternoon.

John

-----------------------------------
Professor John Kitchin
Doherty Hall A207F
Department of Chemical Engineering
Carnegie Mellon University
Pittsburgh, PA 15213
412-268-7803
@johnkitchin
http://kitchingroup.cheme.cmu.edu

**Comment by jkitchin** [comment]
I haven't been able to reproduce this issue this morning so far with either the MELPA or git version. Any chance it has cleared itself up?

**Comment by brittAnderson** [comment]
What has been cleared up is the issue with my brain and not with your
code.

For reasons that I find interesting, if inexplicable, I simply started
to use the command (doi-utils-insert-bibtex-entry-from-doi DOI)

instead of:

(doi-utils-add-bibtex-entry-from-doi DOI BIBFILE)

Note **insert** instead of **add**

I don't know why I did this, since I had been using org-ref for days. I glitched, and I just didn't see the error in my usage (perhaps because
helm prompts "hid it" from me given the short cuts I was using). Anyway,
I finally saw in a helm buffer that both forms were there and tried the
"add" and all works fine. It might be worth putting a condition in the insert function in case someone else confuses inserting with adding. As you get more users, someone else could have the same problem.

In any case, the basic functionality is all there. But back to the original issue, I don't think the generic function is  sufficiently generic. I
tried to experiment with mapping the *doi-redirect* stuff over ".pdf"
and ".full.pdf", but the url-http-file-exits-p returned t for all of
them, even if only one linked to the pdf, and the other linked to pages
saying there was no pdf. So, I don't actually understand how that predicate test works. I have been wanting to learn some elisp and I can look into this in the future.

With a better predicate for whether the file exists, I think we could
use the -map and -filter commands from the dash library to write a nice
concise function to create the various combinations of urls from the
doi redirects and find the one that actually links to the paper.

But that can be some future work for me, and you can close this issue if you want to, because I need a break (apparently). Thanks for your quick support, and I apologize for any time of yours I have wasted. 
In summary, it was a PICNIC problem (problem in chair not in computer)

**Comment by jkitchin** [comment]
Good to hear. I made the function causing you problems not interactive so it won't show up for users to use. The best way to test for a link so far is to simply download it, and if you get a file check that it starts with the magic pdf (see https://github.com/jkitchin/org-ref/blob/master/doi-utils.el#L465). This is not too fast, and so that should be a last resort method. I guess the simple generic function work isn't effective for places that return an http status that doesn't indicate an error.
